src/install/lockfile.zig

```diff
@@ -821,15 +824,51 @@ pub fn clean(
 }
 
 /// Is this a direct dependency of the workspace root package.json?
-pub fn isWorkspaceRootDependency(this: *Lockfile, id: DependencyID) bool {
+pub fn isWorkspaceRootDependency(this: *const Lockfile, id: DependencyID) bool {
     return this.packages.items(.dependencies)[0].contains(id);
 }
 
 /// Is this a direct dependency of the workspace the install is taking place in?
-pub fn isRootDependency(this: *Lockfile, manager: *PackageManager, id: DependencyID) bool {
+pub fn isRootDependency(this: *const Lockfile, manager: *PackageManager, id: DependencyID) bool {
     return this.packages.items(.dependencies)[manager.root_package_id.get(this, manager.workspace_name_hash)].contains(id);
 }
 
+/// Is this a direct dependency of any workspace (including workspace root)?
+/// TODO make this faster by caching the workspace package ids
+pub fn isWorkspaceDependency(this: *const Lockfile, id: DependencyID) bool {
+    const packages = this.packages.slice();
+    const resolutions = packages.items(.resolution);
+    const dependencies_lists = packages.items(.dependencies);
+    for (resolutions, dependencies_lists) |resolution, dependencies| {
+        if (resolution.tag != .workspace and resolution.tag != .root) continue;
+        if (dependencies.contains(id)) return true;
+    }
+
+    return false;
+}
+
+/// Does this tree id belong to a workspace (including workspace root)?
+pub fn isWorkspaceTreeId(this: *const Lockfile, id: Tree.Id) bool {
+    return id == 0 or this.buffers.dependencies.items[this.buffers.trees.items[id].dependency_id].behavior.isWorkspaceOnly();
+}
+
+pub fn getWorkspacePackageID(this: *const Lockfile, workspace_name_hash: ?PackageNameHash) PackageID {
+    return if (workspace_name_hash) |workspace_name_hash_| brk: {
+        const packages = this.packages.slice();
+        const name_hashes = packages.items(.name_hash);
+        const resolutions = packages.items(.resolution);
+        const metas = packages.items(.meta);
+        for (resolutions, name_hashes, metas) |res, name_hash, meta| {
```

Comment:
We can avoid looping through `metas`

Package IDs are sequential, so the index of the current element is also the package id in meta.

We shouldn't really store the package IDs in meta, but we do for historical reasons.